src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+use Tensor\Matrix;
+use Tensor\Vector;
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Word2Vec
+ *
+ * A shallow, two-layer neural network, that produces word embeddings used in NLP models.
+ * This implementation utilizes the skip-gram algorithm and hierarchical softmax or negative sampling.
+ *
+ * References:
+ * [1] `Tomas Mikolov et al: Efficient Estimation of Word Representations
+ * in Vector Space <https://arxiv.org/pdf/1301.3781.pdf>`_, `Tomas Mikolov et al: Distributed Representations of Words
+ * and Phrases and their Compositionality <https://arxiv.org/abs/1310.4546>`
+ *
+ * @category    Machine Learning
+ * @package     RubixML
+ * @author      Rich Davis
+ */
+class Word2Vec implements Embedder, Stateful
+{
+    /**
+     * Presetting random multiplier used to validate word probabilities in sub sampling.
+     *
+     * @var int
+     */
+    protected const RAND_MULTIPLIER = 4294967296;
+
+    /**
+     * The minimum allowed alpha while training.
+     *
+     * @var float
+     */
+    protected const MIN_ALPHA = 0.0001;
+
+    /**
+     * The negative sampling exponent.
+     *
+     * @var float
+     */
+    protected const NS_EXPONENT = 0.75;
+
+    /**
+     * An array of sanitized and exploded sentences.
+     *
+     * @var array[]
+     */
+    protected $corpus = [];
+
+    /**
+     * An array of each word in the corpus for preprocessing purposes.
+     *
+     * @var int[]
+     */
+    protected $rawVocab = [];
+
+    /**
+     * An array containing each word in the corpus and it's respective index, count, and multiplier.
+     *
+     * @var mixed[]
+     */
+    protected $vocab = [];
+
+    /**
+     * An array containing each word in the corpus at it's respective index.
+     *
+     * @var int[]
+     */
+    protected $index2word = [];
+
+    /**
+     * An array containing the hidden layer.
+     *
+     * @var Vector[]
+     */
+    protected $syn1 = [];
+
+    /**
+     * An array containing word vectors.
+     *
+     * @var Vector[]
+     */
+    protected $vectors = [];
+
+    /**
+     * The error vector used during training.
+     *
+     * @var \Tensor\Vector
+     */
+    protected $error;
+
+    /**
+     * The lock factors for each word in the context.
+     *
+     * @var int[]
+     */
+    protected $vectorsLockf = [];
+
+    /**
+     * The L2-normalized word vectors from the model.
+     *
+     * @var Vector[]
+     */
+    protected $vectorsNorm = [];
+
+    /**
+     * The cumulative distrubtion table for negative sampling.
+     *
+     * @var int[]
+     */
+    protected $cumTable = [];
+
+    /**
+     * The last digit in the cumulative distribution table.
+     *
+     * @var int
+     */
+    protected $endCumDigit;
+
+    /**
+     * The negative labels used in the cumulative distrubtion table.
+     *
+     * @var \Tensor\Vector
+     */
+    protected $negLabels;
+
+    /**
+     * The training method determined by the layer they determine.
```

Comment:
Can you clarify this description? I having trouble understanding what you are trying to convey. Who/what is they?

Reply:
This is whether they chose hierarchal softmax or negative sampling.

Reply:
Can you communicate that in this comment for the user?